[FLINK-15640][k8s] Support to set labels for jobmanager and taskmanager pod

wangyang0918 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Introduce config options to support to set labels and node-selector for JM/TM pods. 
* The [labels](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/) are key/value pairs that will be attached to the JM/TM pod.


## Brief change log

* Support to set labels

## Verifying this change

* Unit tests
* Manually test in a real cluster

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (yes)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (docs)

wangyang0918 commented:
@TisonKun @zhengcanbin Do you mind to take a look this PR?

zhengcanbin commented:
Hi, @wangyang0918, this PR introduces two features, how about splitting it into two separate PRs?

wangyang0918 commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/configuration/KubernetesConfigOptions.java:
```diff
@@ -111,6 +112,34 @@
+			"For example, version:alphav1,deploy:test.");
+
+	public static final ConfigOption<Map<String, String>> JOB_MANAGER_NODE_SELECTOR =
+		key("kubernetes.jobmanager.node-selector")
```
I use singular for `node-selector` because in the K8s yaml file it is always singular. For `labels` and `annotations`, they are plural. So i am just trying to keep the same behavior with K8s document.
If you insist of changing it to `node-selectors`, i could update it.

zhengcanbin replied:
> I use singular for `node-selector` because in the K8s yaml file it is always singular. For `labels` and `annotations`, they are plural. So i am just trying to keep the same behavior with K8s document.
> If you insist of changing it to `node-selectors`, i could update it.

Indeed, the K8s use the singular for node-selector. I think you are right, we'd better try our best to keep consistent with the K8s side.

wangyang0918 commented on flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/parameters/KubernetesJobManagerParameters.java:
```diff
@@ -52,10 +54,16 @@ public KubernetesJobManagerParameters(Configuration flinkConfig, ClusterSpecific
-		Map<String, String> labels = getCommonLabels();
+		final Map<String, String> labels = new HashMap<>(getCommonLabels());
 		labels.put(Constants.LABEL_COMPONENT_KEY, Constants.LABEL_COMPONENT_JOB_MANAGER);
+		flinkConfig.getOptional(KubernetesConfigOptions.JOB_MANAGER_LABELS).ifPresent(labels::putAll);
```
I think you are right. I will prioritize the built-in labels.

wangyang0918 commented:
@zhengcanbin Thanks for you review. I think your suggestion is on point. I will split this ticket into two separate ones.

wangyang0918 commented:
@zhengcanbin I have addressed your comments. Could you take a look at your convenience?

@TisonKun Could you also have a final check and help to merge?

zhengcanbin commented:
@wangyang0918 It would be good to rebase this PR on the latest master and resolve the merge conflicts.

wangyang0918 commented:
@zhengcanbin Thanks for your review. I have rebased on the master.

@TisonKun Could you please also have a look?
